File: src/com/dugonggames/elclasico/DistanceClassifier.java
package com.dugonggames.elclasico;

public class DistanceClassifier {
    DigitImage[] avgs;

    public DistanceClassifier(DigitImage[] images, int[] labels){
        avgs = new DigitImage[10];
        for (int i = 0; i < 10; i++) avgs[i] = new DigitImage(new float[784], 0);
        int[] counts = new int[10];

        for (int i = 0; i < 10000; i++){
            avgs[labels[i]] = avgs[labels[i]].add(images[i]);
            counts[labels[i]]++;
        }

        for (int i = 0; i < 10; i++) avgs[i] = avgs[i].divide(counts[i]);
    }

    public int eClassify(DigitImage image){
        double smallestDist = Double.POSITIVE_INFINITY;
        int smallestDistIndex= -1;
        for (int i = 0; i < 10; i++){
            if (image.eDistance(avgs[i]) < smallestDist){
                smallestDist = image.eDistance(avgs[i]);
                smallestDistIndex = i;
            }
        }
        return smallestDistIndex;
    }

    public int tClassify(DigitImage image){
        double smallestDist = Double.POSITIVE_INFINITY;
        int smallestDistIndex= -1;
        for (int i = 0; i < 10; i++){
            if (image.tDistance(avgs[i]) < smallestDist){
                smallestDist = image.tDistance(avgs[i]);
                smallestDistIndex = i;
            }
        }
        return smallestDistIndex;
    }
}

File: src/com/dugonggames/elclasico/DistanceClassifier.kt
package com.dugonggames.elclasico

class DistanceClassifier(images: Array<DigitImage>, labels: IntArray) {
    internal var avgs: Array<DigitImage>

    init {
        avgs = Array(10) { DigitImage(FloatArray(784), 0) }
        val counts = IntArray(10)

        for (i in 0..9999) {
            avgs[labels[i]] = avgs[labels[i]].add(images[i])
            counts[labels[i]]++
        }

        for (i in 0..9) avgs[i] = avgs[i].divide(counts[i])
    }

    fun eClassify(image: DigitImage): Int {
        var smallestDist = java.lang.Double.POSITIVE_INFINITY
        var smallestDistIndex = -1
        for (i in 0..9) {
            if (image.eDistance(avgs[i]) < smallestDist) {
                smallestDist = image.eDistance(avgs[i])
                smallestDistIndex = i
            }
        }
        return smallestDistIndex
    }

    fun tClassify(image: DigitImage): Int {
        var smallestDist = java.lang.Double.POSITIVE_INFINITY
        var smallestDistIndex = -1
        for (i in 0..9) {
            if (image.tDistance(avgs[i]) < smallestDist) {
                smallestDist = image.tDistance(avgs[i])
                smallestDistIndex = i
            }
        }
        return smallestDistIndex
    }
}
